### PR TITLE
[SPARK-32276][SQL] Remove redundant sorts before repartition nodes

### PR DESCRIPTION
**Pull request by aokolnychyi**
### What changes were proposed in this pull request?

This PR proposes to remove redundant sorts before repartition nodes whenever the data is ordered after the repartitioning. 

### Why are the changes needed?

It looks like our `EliminateSorts` rule can be extended further to remove sorts before repartition nodes that don't affect the final output ordering. It seems safe to perform the following rewrites:

- `Sort -> Repartition -> Sort -> Scan` as `Sort -> Repartition -> Scan`
- `Sort -> Repartition -> Project -> Sort -> Scan` as `Sort -> Repartition -> Project -> Scan`

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

More test cases.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -981,6 +982,10 @@ object EliminateSorts extends Rule[LogicalPlan] {
       j.copy(left = recursiveRemoveSort(originLeft), right = recursiveRemoveSort(originRight))
     case g @ Aggregate(_, aggs, originChild) if isOrderIrrelevantAggs(aggs) =>
       g.copy(child = recursiveRemoveSort(originChild))
+    case r: RepartitionByExpression =>
```
Comment (aokolnychyi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:985]: These two branches can be replaces with one:

```
case r: RepartitionOperation =>
  r.withNewChildren(r.children.map(recursiveRemoveSort))
```

It will mean all repartition nodes we add in the future will be also taken into account. It seems safe but I want to hear what everybody thinks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala [hunk 2]
```diff
@@ -1299,8 +1299,9 @@ class ArrowConvertersSuite extends SharedSparkSession {
          |}
        """.stripMargin
 
-    val a_i = List[Int](1, -1, 2, -2, 2147483647, -2147483648)
-    val b_i = List[Option[Int]](Some(1), None, None, Some(-2), None, Some(-2147483648))
+    // we order values by $"a_i".desc manually as sortBy before coalesce is ignored
```
Comment (aokolnychyi) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala:1302]: This is the case I mention in the PR description. In this case, `DefaultPartitionCoalescer` does preserve the ordering and the test relied on that even though there is no guarantee it will happen. We could apply the new optimization only if the repartition operation requires a shuffle. That way, we will keep the existing behavior.

**Comment by aokolnychyi** [comment]
cc @dongjoon-hyun @dbtsai @cloud-fan @viirya @gengliangwang for feedback

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @aokolnychyi .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -981,6 +982,10 @@ object EliminateSorts extends Rule[LogicalPlan] {
...
+      r.copy(child = recursiveRemoveSort(r.child))
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:986]: Should we also limit to `partitionExpressions` are all deterministic case?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala (hunk 2 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala:1302]: I think we should apply the optimization only if the repartition requires a shuffle as you suggest. I know that there are users that depend on this behavior.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:986]: I don't think that is necessary. My understanding is that the requirement for non-deterministic expressions is that they are run only once when finally executed, not that the plan cannot be optimized around them.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:985]: What about adding `RepartitionOperation.preservesOrder`? Then we could collapse these cases while also excluding `Coalesce` and making this explicit for future repartition operators.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala [hunk 4]
```diff
@@ -310,4 +310,109 @@ class EliminateSortsSuite extends PlanTest {
     val correctAnswer = PushDownOptimizer.execute(noOrderByPlan.analyze)
     comparePlans(optimized, correctAnswer)
   }
+
+  testRepartitionOptimization(
```
Comment (rdblue) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala:314]: I prefer not to replace `test` because it makes tests difficult to run individually (at least in my IntelliJ environment).

It also tends to increase readability. Here, you're passing a function to `testRepartitionOptimization` that gets passed a function that modifies the logical plan. I think it would be easier to read if these were separate suites, with a suite-level repartition function:

```scala
class EliminateSortsInRepartitionSuite extends ... {
  def repartition(plan: LogicalPlan): LogicalPlan = plan.repartition(10)

  test("remove sortBy") {
    val plan = testRelation.select('a, 'b).sortBy('a.asc, 'b.desc)
    val planWithRepartition = repartition(plan)
    ...
  }
}

class EliminateSortsInRepartitionByExpressionSuite extends EliminateSortsInRepartitionSuite {
  override def repartition(plan: LogicalPlan): LogicalPlan = plan.distribute('a, 'b)(10)
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:986]: Non-deterministic expressions could have status and row-order sensitive. In `EliminateSorts` we also only remove Sort for Project and Filter where projection list and condition are deterministic.

Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:986]: You're right, that does seem to imply that ordering is preserved underneath non-deterministic expressions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (aokolnychyi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:985]: I am +1 on having a repartition node that preserves ordering. In fact, we have such a node internally. `Coalesce` is not really order preserving, though. It has custom logic in `DefaultPartitionCoalescer` that gets applied if the parent RDD has locality info. We would also need to report `outputOrdering` correctly (which is not done in case of `Coalesce` now)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala (hunk 2 above)
Comment (aokolnychyi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala:1302]: I lean towards that as well.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (aokolnychyi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:986]: Agree, let me update this.

Comment (aokolnychyi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:986]: Updated.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+import org.apache.spark.sql.catalyst.catalog.{InMemoryCatalog, SessionCatalog}
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.plans.PlanTest
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.RuleExecutor
+
+class EliminateSortsBeforeRepartitionSuite extends PlanTest {
+
+  val catalog = new SessionCatalog(new InMemoryCatalog, EmptyFunctionRegistry, conf)
```
Comment (aokolnychyi) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala:30]: I considered making these protected but it makes lines longer and most of tests use public vars as well.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (aokolnychyi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We won't be able to squash all cases into one as we need to check if repartition expressions are deterministic. However, I'd consider extending repartition nodes with order preserving repartition in a follow-up if there is enough support for that.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala (hunk 2 above)
Comment (aokolnychyi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala]: Updated.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala (hunk 4 above)
Comment (aokolnychyi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala]: I think this pattern is common for the codebase but I agree having separate suites makes more sense here. Updated.

**Comment by aokolnychyi** [comment]
@dongjoon-hyun @viirya @rdblue this PR is ready for another review round.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        FoldablePropagation,
+        LimitPushDown) ::
+      Batch("Eliminate Sorts", Once,
+        EliminateSorts) ::
+      Batch("Collapse Project", Once,
+        CollapseProject) :: Nil
+  }
+
+  def repartition(plan: LogicalPlan): LogicalPlan = plan.repartition(10)
+  def isOptimized: Boolean = true
```
Comment (rdblue) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala:46]: What is the purpose of this? It is always set to true, so I don't see why the test cases use it.

**Comment by rdblue** [comment]
LGTM. I just had one question in the tests.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala (hunk 6 above)
Comment (aokolnychyi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala:46]: It is set to false in `EliminateSortsBeforeCoalesceSuite` as we excluded coalesce.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+  test(s"sortBy") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala:48]: nit. `s"` -> `"`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    val optimizedPlan = Optimize.execute(analyzer.execute(planWithRepartition))
+    val correctPlan = if (isOptimized) {
+      repartition(testRelation.select('a, 'b))
+    } else {
+      planWithRepartition
+    }
+    comparePlans(optimizedPlan, analyzer.execute(correctPlan))
+  }
+
+  test(s"sortBy with projection") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala:60]: ditto.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala]: Could you update the PR description accordingly?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:989]: Could you update the PR description accordingly?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala (hunk 6 above)
Comment (rdblue) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala:46]: I see. I missed that suite. Thanks for explaining!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (aokolnychyi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:989]: Done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala (hunk 7 above)
Comment (aokolnychyi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala]: Good catch, these are leftovers from the old version of tests.

Comment (aokolnychyi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala]: Fixed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala (hunk 8 above)
Comment (aokolnychyi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsBeforeRepartitionSuite.scala]: Fixed.

**Comment by viirya** [comment]
Some tests in `EliminateSortsSuite` were failed?

**Comment by aokolnychyi** [comment]
Those failures seem to belong to the old commit before [this](https://github.com/apache/spark/pull/29089/commits/c58ad12609b732fe8539277f3dcfb046268c5740) change where I removed old tests. Tests for coalesce that failed were no longer valid.

**Review by viirya — APPROVED** [review]

**Review by maropu — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @aokolnychyi , @rdblue , @viirya , @maropu .
Merged to master for Apache Spark 3.1.0 on December 2020.
The tests already passed. Last two comments are about commenting and changing test case names.

**Comment by dongjoon-hyun** [comment]
Also, cc @gatorsmile and @cloud-fan

**Comment by aokolnychyi** [comment]
Thanks, everyone!

**Comment by dongjoon-hyun** [comment]
Oops. Sorry, guys. It seems that I missed something during testing. For the following case, we should not remove `Sort`.

**BEFORE THIS PR**
```scala
scala> Seq((1,10),(1,20),(2,30),(2,40)).toDF("a", "b").repartition(2).createOrReplaceTempView("t")

scala> sql("select * from (select * from t order by b desc) distribute by a").show()
+---+---+
|  a|  b|
+---+---+
|  1| 20|
|  1| 10|
|  2| 40|
|  2| 30|
+---+---+
```

**AFTER THIS PR**
```scala
scala> Seq((1,10),(1,20),(2,30),(2,40)).toDF("a", "b").repartition(2).createOrReplaceTempView("t")

scala> sql("select * from (select * from t order by b desc) distribute by a").show()
+---+---+
|  a|  b|
+---+---+
|  1| 10|
|  1| 20|
|  2| 30|
|  2| 40|
+---+---+
```

**Comment by dongjoon-hyun** [comment]
To generate small Parquet/ORC files, we do the above tricks, don't we?

**Comment by dongjoon-hyun** [comment]
**AFTER SPARK-32276**
```
scala> scala.util.Random.shuffle((1 to 100000).map(x => (x % 2, x))).toDF("a", "b").repartition(2).createOrReplaceTempView("t")

scala> sql("select * from (select * from t order by b) distribute by a").write.orc("/tmp/SPARK-32276")
```

```
$ ls -al /tmp/SPARK-32276/
total 632
drwxr-xr-x  10 dongjoon  wheel     320 Jul 14 22:08 ./
drwxrwxrwt  14 root      wheel     448 Jul 14 22:08 ../
-rw-r--r--   1 dongjoon  wheel       8 Jul 14 22:08 ._SUCCESS.crc
-rw-r--r--   1 dongjoon  wheel      12 Jul 14 22:08 .part-00000-ba5049f9-b835-49b7-9fdb-bdd11b9891cb-c000.snappy.orc.crc
-rw-r--r--   1 dongjoon  wheel    1188 Jul 14 22:08 .part-00043-ba5049f9-b835-49b7-9fdb-bdd11b9891cb-c000.snappy.orc.crc
-rw-r--r--   1 dongjoon  wheel    1188 Jul 14 22:08 .part-00191-ba5049f9-b835-49b7-9fdb-bdd11b9891cb-c000.snappy.orc.crc
-rw-r--r--   1 dongjoon  wheel       0 Jul 14 22:08 _SUCCESS
-rw-r--r--   1 dongjoon  wheel     119 Jul 14 22:08 part-00000-ba5049f9-b835-49b7-9fdb-bdd11b9891cb-c000.snappy.orc
-rw-r--r--   1 dongjoon  wheel  150735 Jul 14 22:08 part-00043-ba5049f9-b835-49b7-9fdb-bdd11b9891cb-c000.snappy.orc
-rw-r--r--   1 dongjoon  wheel  150741 Jul 14 22:08 part-00191-ba5049f9-b835-49b7-9fdb-bdd11b9891cb-c000.snappy.orc
```

**BEFORE**
```
scala> scala.util.Random.shuffle((1 to 100000).map(x => (x % 2, x))).toDF("a", "b").repartition(2).createOrReplaceTempView("t")

scala> sql("select * from (select * from t order by b) distribute by a").write.orc("/tmp/master")
```

```
$ ls -al /tmp/master/
total 56
drwxr-xr-x  10 dongjoon  wheel  320 Jul 14 22:12 ./
drwxrwxrwt  15 root      wheel  480 Jul 14 22:12 ../
-rw-r--r--   1 dongjoon  wheel    8 Jul 14 22:12 ._SUCCESS.crc
-rw-r--r--   1 dongjoon  wheel   12 Jul 14 22:12 .part-00000-2cd3a50e-eded-49a4-b7cf-94e3f090b8c1-c000.snappy.orc.crc
-rw-r--r--   1 dongjoon  wheel   16 Jul 14 22:12 .part-00043-2cd3a50e-eded-49a4-b7cf-94e3f090b8c1-c000.snappy.orc.crc
-rw-r--r--   1 dongjoon  wheel   16 Jul 14 22:12 .part-00191-2cd3a50e-eded-49a4-b7cf-94e3f090b8c1-c000.snappy.orc.crc
-rw-r--r--   1 dongjoon  wheel    0 Jul 14 22:12 _SUCCESS
-rw-r--r--   1 dongjoon  wheel  119 Jul 14 22:12 part-00000-2cd3a50e-eded-49a4-b7cf-94e3f090b8c1-c000.snappy.orc
-rw-r--r--   1 dongjoon  wheel  932 Jul 14 22:12 part-00043-2cd3a50e-eded-49a4-b7cf-94e3f090b8c1-c000.snappy.orc
-rw-r--r--   1 dongjoon  wheel  939 Jul 14 22:12 part-00191-2cd3a50e-eded-49a4-b7cf-94e3f090b8c1-c000.snappy.orc
```

**Comment by dongjoon-hyun** [comment]
Very sorry, guys. Due to the above regression, I'll revert this commit urgently. We can rethink about this PR.

**Comment by dongjoon-hyun** [comment]
I made a PR to add a test coverage for the above case.
- https://github.com/apache/spark/pull/29118

**Comment by viirya** [comment]
Oh, this is interesting. I know removing `Sort` before `Repartition` will result in different data distribution because `Repartition` uses `RoundRobinPartitioning`. Because I think repartition doesn't guarantee shuffled data distribution, so I thought it is okay.

Now seems different data distribution causes difference storage output size. I think it is because to repartition sorted data using `RoundRobinPartitioning` can generate more compact output.

**Comment by dongjoon-hyun** [comment]
No~ It depends on file formats instead of Spark side.
For example, in the above example, ORC files are small because it supports a special encoding when the data is sorted with a fixed increment. For Parquet files, the result will be different.

**Comment by viirya** [comment]
Yeah, because the different data distribution, physical encoding of data could result in different size.

**Comment by hvanhovell** [comment]
Ehh... AFAIK nested ordering can be ignored from a relation algebra point of view. So I am not sure this is a very solid argument. This feels a bit like an example of [hyrum's law](https://www.hyrumslaw.com/). If you want sorted runs in ORC then you ought to fix is there, and not rely on some implicit system behavior.

Regarding the shuffles. If the data is sorted before it goes into the shuffle, then the individual shuffle blocks are sorted. This is also the reason why doing a sort aggregate is not completely terrible (TimSort is good at identifying sorted runs).

**Comment by dongjoon-hyun** [comment]
@hvanhovell . I agree with you for the followings.
> AFAIK nested ordering can be ignored from a relation algebra point of view. 
> Regarding the shuffles. ...

However, the following is not reasonable. There is nothing wrong in the file formats. They are just consumers and showing a better performance in a sorted input sequence. I guess you assume that this is only a behavior at ORC. But, I'm sure that you can find your customers are relying on this in Parquet, too.
> If you want sorted runs in ORC then you ought to fix is there, and not rely on some implicit system behavior.

This is not an implicit system behavior in Apache Spark. Apache Spark has been working in the procedural ways as you see in the above. If we start to ignore the valid working pattern in the production, it becomes a huge regression.

In short, `saving` to a file is a totally different and valid story. To optimize the final output files, the above pattern have been used in the production among Apache Spark users for a long time. If some optimizer rule ignores the existing usage, this ends up at a large regression in terms of the cost (for example, S3) obviously.

**Comment by dongjoon-hyun** [comment]
cc @marmbrus and @gatorsmile since they know the existing customers well and are good at protecting them.

**Comment by dongjoon-hyun** [comment]
BTW, @aokolnychyi . I merged the corner case test case. Could you rebase this to the master? Then, we can discuss how to proceed this PR in a narrowed direction.

**Comment by aokolnychyi** [comment]
Oops, thanks for catching the corner case quickly, @dongjoon-hyun!

My original idea for this PR was based on the fact that a range partitioning followed by a local sort is equivalent to a global sort if expressions are compatible. Then I started to generalize this idea and there was no obvious corner case. While this one is very subtle, I think it makes sense if we think more about it. Repartition nodes change data distribution but may not necessarily change the ordering of data. Partially, this is the reason why we excluded coalesce in the original proposal. Based on the example above, this seems to be true even if we hash partition our data.

I'd explore cases where sort+repartition are next to each other. In that case, we are sure we change both the ordering and distribution and can potentially ignore the sort below. 

For example, we may have this:

```
sql("select * from (select * from (select * from t order by b desc) distribute by a) order by b asc")
```

```
Sort [b#6 ASC NULLS FIRST], true
+- RepartitionByExpression [a#5], 4
   +- Sort [b#6 DESC NULLS LAST], true
      +- Repartition 2, true
         +- LocalRelation [a#5, b#6]
```

Is there a case where we want to keep the first sort before RepartitionByExpression?

**Comment by aokolnychyi** [comment]
The same question for local sort too.

```
sql("select * from (select * from (select * from t order by b desc) distribute by a) sort by b asc")
```

```
Sort [b#6 ASC NULLS FIRST], false
+- RepartitionByExpression [a#5], 4
   +- Sort [b#6 DESC NULLS LAST], true
      +- Repartition 2, true
         +- LocalRelation [a#5, b#6]
```

**Comment by cloud-fan** [comment]
In general, I think we can remove sort if it doesn't affect the final output ordering. The case caught by @dongjoon-hyun is a good example: the final output ordering changes and affect the file size.

**Comment by rdblue** [comment]
> To generate small final Parquet/ORC files, we do the above tricks, don't we?

We don't rely on this. Our recommendation to users is to add a global sort to distribute the data, which adds the local sort in the final stage that won't be removed. I can understand people relying on this behavior, though.

For now, I think it makes sense to remove a sort before a repartition if the data will be sorted later, like what I think @aokolnychyi is suggesting. That's really what we will need for tables that require a sort order -- that will be the final sort and we should be able to remove other sorts.

We may also want to choose whether this is a guarantee and document it.

**Comment by aokolnychyi** [comment]
Yes, my proposal is to optimize cases when we sort the data after the repartition like in the examples I gave above. In those cases, sorts below seem to be redundant.

**Comment by aokolnychyi** [comment]
@dongjoon-hyun @viirya @hvanhovell @maropu, what do you think?

**Comment by viirya** [comment]
It looks reasonable to me to remove a sort before a repartition if we know the data will be sorted later, e.g. @aokolnychyi's examples above.

**Comment by aokolnychyi** [comment]
I've updated the PR to show what I meant. I'll check for additional edge cases in the morning but the change is ready for review.

**Comment by dongjoon-hyun** [comment]
Thank you for quick updating, @aokolnychyi . Also, thank you all for your opinions.

**Comment by aokolnychyi** [comment]
I gave it a bit of thought and did not find a case where the updated logic would break.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 9]
```diff
@@ -969,11 +969,11 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
...
- * 3) if there is another Sort operator separated by 0...n Project/Filter operators
- * 4) if the Sort operator is within Join separated by 0...n Project/Filter operators only,
- *    and the Join conditions is deterministic
- * 5) if the Sort operator is within GroupBy separated by 0...n Project/Filter operators only,
- *    and the aggregate function is order irrelevant
+ * 3) if there is another Sort operator separated by 0...n Project/Filter/Repartition operators
+ * 4) if the Sort operator is within Join separated by 0...n Project/Filter/Repartition
+ *    operators only, and the Join conditions is deterministic
+ * 5) if the Sort operator is within GroupBy separated by 0...n Project/Filter/Repartition
+ *    operators only, and the aggregate function is order irrelevant
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:976]: This documentation update seems to focus on `case _: Repartition => true` only. Could you revise more to cover `case r: RepartitionByExpression => r.partitionExpressions.forall(_.deterministic)`, please?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 10]
```diff
@@ -969,11 +969,11 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
  * Removes Sort operation. This can happen:
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:969]: Shall we revise this line?
```scala
- Removes Sort operation. This can happen:
+ Removes Sort operation if it doesn't affect the final output ordering.
+ Note that the final output ordering changes and affect the file size (SPARK-32318).
+ This optimizer handles the following cases:
```

Comment (aokolnychyi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 9 above)
Comment (aokolnychyi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Done.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you so much, @aokolnychyi and all.
The contribution of this PR is not only on improving the optimizer but also adding more test coverage.
Merged to master.

**Comment by dongjoon-hyun** [comment]
cc @cloud-fan and @gatorsmile once more.

**Comment by hvanhovell** [comment]
@dongjoon-hyun I am a bit late with my response but here goes :)

> However, the following is not reasonable. There is nothing wrong in the file formats. They are just consumers and showing a better performance in a sorted input sequence because they are columnar vectorized format. I guess you assume that this is only a behavior at ORC. But, I'm sure that you can find your customers are relying on this in Parquet, too.

That is making the argument for explicitly organizing the data before the write right? You are currently just lucky that the system accidentally produces a nice layout for you; 99% of our users won't be as lucky. The only way you can he sure, is when you add these things yourself.

> This is not an implicit system behavior in Apache Spark. Apache Spark has been working in the procedural ways as you see in the above. If we start to ignore the valid working pattern in the production, it becomes a huge regression.
> In short, saving to a file is a totally different and valid story. To optimize the final output files, the above pattern have been used in the production among Apache Spark users for a long time. If some optimizer rule ignores the existing usage, this ends up at a large regression in terms of the cost (for example, S3) obviously.

If you generalize the procedural argument then we also should not do things like join reordering or swapping window operators. The whole point of a declarative system like Spark SQL is that you don't care about how the system executes a query, and that it has the freedom move operations around to make execution more optimal.

Have you considered that your regression is someone else's speed-up? Sorting is not free, and if we can avoid it we should. There might a large group of users that are adversely affected by spurious sorts in the queries (e.g. an order by in a view).

Finally I do want to point out that there is no mechanism that captures this regression if it pops up again.

**Comment by dongjoon-hyun** [comment]
@hvanhovell . The following is complete wrong because the above optimization was one of the recommendations for many Hortonworks customers to save their HDFS usage. I knew many production usages like that. I almost forgot that, but it rang my head suddenly during this PR. (Sadly, after I merged this.)
>  You are currently just lucky that the system accidentally produces a nice layout for you; 99% of our users won't be as lucky. The only way you can he sure, is when you add these things yourself.

I understand your point of views fully. However, I'm wondering if you can persuade the customer to waste their storage by generating 160x bigger files (SPARK-32318). Do you think you can?

```
-rw-r--r--   1 dongjoon  wheel  939 Jul 14 22:12 part-00191-2cd3a50e-eded-49a4-b7cf-94e3f090b8c1-c000.snappy.orc
```
```
-rw-r--r--   1 dongjoon  wheel  150741 Jul 14 22:08 part-00191-ba5049f9-b835-49b7-9fdb-bdd11b9891cb-c000.snappy.orc
```

**Comment by dongjoon-hyun** [comment]
For the following, I added SPARK-32318 added a test coverage at master/3.0/2.4. Are you suggesting that's not enough?
> Finally I do want to point out that there is no mechanism that captures this regression if it pops up again.

**Comment by dongjoon-hyun** [comment]
For the following, I'd like to ask your help if you are interested. I believe we want to build the better Apache Spark in the community together.
> If you generalize the procedural argument then we also should not do things like join reordering or swapping window operators.
